### ISSUE TITLE
How evaluate the NN trained with own created dataset in COCO-Style?

### ISSUE THREAD

**Issue by thepate94227** [matterport/Mask_RCNN#1466]
I created my own dataset in COCO-style. So i have the annotation-files for my trainset, validationset, testset and for all my images. 
I can use the prediction to see how the Ground Truth looks like and what my NN predicts:

```
dataset_test = coco.CocoDataset()
dataset_test.load_coco(DATA_DIR, subset="test", year="2019")
dataset_test.prepare()

model.load_weights(model_path, by_name=True)

image_id = random.choice(dataset_test.image_ids)
original_image, image_meta, gt_class_id, gt_bbox, gt_mask =\
    modellib.load_image_gt(dataset_test, inference_config,
                           image_id, use_mini_mask=False)

log("original_image", original_image)
log("image_meta", image_meta)
log("gt_class_id", gt_class_id)
log("gt_bbox", gt_bbox)
log("gt_mask", gt_mask)
print("image_id: ", image_id)

visualize.display_instances(original_image, gt_bbox, gt_mask, gt_class_id,
                            dataset_train.class_names, figsize=(8, 8))

results = model.detect([original_image], verbose=1)

r = results[0]

visualize.display_instances(original_image, r['rois'], r['masks'], r['class_ids'],
                            dataset_test.class_names, r['scores'], figsize=(8, 8))
```
But how can i go through all my testfiles to see how good my NN really is?
I already have all the GTs for my images and I trained my NN for a very long time. But the prediction is bad. I want to see if the NN "thinks" it is good or if the NN is just bad.

I have nearly 600 train images and 120 validation images. Here is an example for GT and prediction with the config:

```
    GPU_COUNT = 1
    IMAGES_PER_GPU = 2
    NUM_CLASSES = 1 + 1  # background + 1 rope
    #BACKBONE = "resnet101"
    BACKBONE = "resnet50"
    # Aim to allow ROI sampling to pick 33% positive ROIs.
    TRAIN_ROIS_PER_IMAGE = 32
    STEPS_PER_EPOCH = 582
    VALIDATION_STEPS = STEPS_PER_EPOCH / 6
    LEARNING_RATE = 0.001

model.train(dataset_train, dataset_validate, 
            learning_rate=config.LEARNING_RATE / 100,
            epochs=480
            layers="all",
            augmentation=augmentation)
```
![gt222](https://user-images.githubusercontent.com/5870642/56584746-b1b2b400-65dc-11e9-9097-3a56efbd6972.png)
![predict2](https://user-images.githubusercontent.com/5870642/56584757-b5ded180-65dc-11e9-981f-3377ea65700f.png)

The prediction said: There is no instances to display. But obviously in the GT you see, that they are ropes...
Therfore i want to go through all my testimages and evaluate the results. I can't believe that after that many epochs the NN don't recognize any ropes. Even with my trainset the prediction is bad.

**Comment by ehsanpi** [comment]
Specify further on your platform and some of codes e.g. the code for the anchored ropes.
You may use a colour segmentation and get it to count the pixels for you; though not very accurate it well approaches a human recognition since neither of mask creation in the training procedure is 100% accurate.

**Comment by thepate94227** [comment]
I use OpenSuse Leap 15 and PyCharm. 
Which code exactly? I created the images and the Ground Truth myself and then used waspinators tool to transform my images and labels to the COCO format, in .json annotation file. (https://github.com/waspinator/pycococreator)

It seem to works fine, because the visualization of my Ground Truth looks great and correct. I checked many Ground Truth Files, all of them looks correct.

So i have 4 datasets in COCO-format with the .json files:
My Testset, Trainset, Validationset and all my images. I use the dataset which contains all my images later for evaluation, because the coco.py file does the same thing for their COCO-dataset valuation.

The other files like coco.py are from this Mask-RCNN release. 
I thought that Mask-RCNN has his own tools for evaluation. Therfore i asked. 

I think i can evaluate my results with this code:

```
import time
import os
import numpy as np

import mcoco.coco as coco
import mrcnn.model as modellib
from mrcnn.config import Config
from pycocotools.coco import COCO
from pycocotools.cocoeval import COCOeval
from pycocotools import mask as maskUtils


HOME_DIR = 'My-Home-Dir.....'
DATA_DIR = os.path.join(HOME_DIR, "data/knot")
MODEL_DIR = os.path.join(DATA_DIR, "logs")

image_size = 1024
rpn_anchor_template = (1, 2, 4, 8, 16)  # anchor sizes in pixels
rpn_anchor_scales = tuple(i * (image_size // 16) for i in rpn_anchor_template)
class KnotConfig(Config):
    """Configuration for training on the shapes dataset.
    """
    NAME = "knot"
    GPU_COUNT = 1
    IMAGES_PER_GPU = 2
    NUM_CLASSES = 1 + 1  # background + 1 rope
    BACKBONE = "resnet50"
    RPN_ANCHOR_SCALES = rpn_anchor_scales
    TRAIN_ROIS_PER_IMAGE = 32
    STEPS_PER_EPOCH = 582
    VALIDATION_STEPS = STEPS_PER_EPOCH / 6
    USE_MINI_MASK = False
    LEARNING_RATE = 0.001

config = KnotConfig()

def build_coco_results(dataset, image_ids, rois, class_ids, scores, masks):
    """Arrange resutls to match COCO specs in http://cocodataset.org/#format
    """
    # If no results, return an empty list
    if rois is None:
        return []

    results = []
    for image_id in image_ids:
        # Loop through detections
        for i in range(rois.shape[0]):
            class_id = class_ids[i]
            score = scores[i]
            bbox = np.around(rois[i], 1)
            mask = masks[:, :, i]

            result = {
                "image_id": image_id,
                "category_id": dataset.get_source_class_id(class_id, "coco"),
                "bbox": [bbox[1], bbox[0], bbox[3] - bbox[1], bbox[2] - bbox[0]],
                "score": score,
                "segmentation": maskUtils.encode(np.asfortranarray(mask))
            }
            results.append(result)
    return results


def evaluate_coco(model, dataset, coco, eval_type="bbox", limit=0, image_ids=None):
    """Runs official COCO evaluation.
    dataset: A Dataset object with valiadtion data
    eval_type: "bbox" or "segm" for bounding box or segmentation evaluation
    limit: if not 0, it's the number of images to use for evaluation
    """
    # Pick COCO images from the dataset
    image_ids = image_ids or dataset.image_ids

    # Limit to a subset
    if limit:
        image_ids = image_ids[:limit]

    # Get corresponding COCO image IDs.
    coco_image_ids = [dataset.image_info[id]["id"] for id in image_ids]

    t_prediction = 0
    t_start = time.time()

    results = []
    for i, image_id in enumerate(image_ids):
        # Load image
        image = dataset.load_image(image_id)

        # Run detection
        t = time.time()
        r = model.detect([image], verbose=0)[0]
        t_prediction += (time.time() - t)

        # Convert results to COCO format
        # Cast masks to uint8 because COCO tools errors out on bool
        image_results = build_coco_results(dataset, coco_image_ids[i:i + 1],
                                           r["rois"], r["class_ids"],
                                           r["scores"],
                                           r["masks"].astype(np.uint8))
        results.extend(image_results)

    # Load results. This modifies results with additional attributes.
    coco_results = coco.loadRes(results)

    # Evaluate
    cocoEval = COCOeval(coco, coco_results, eval_type)
    cocoEval.params.imgIds = coco_image_ids
    cocoEval.evaluate()
    cocoEval.accumulate()
    cocoEval.summarize()

    print("Prediction time: {}. Average {}/image".format(
        t_prediction, t_prediction / len(image_ids)))
    print("Total time: ", time.time() - t_start)

class InferenceConfig(KnotConfig):
    GPU_COUNT = 1
    IMAGES_PER_GPU = 1
    
inference_config = InferenceConfig()

model = modellib.MaskRCNN(mode="inference",
                          config=inference_config,
                          model_dir=MODEL_DIR)

model_path = model.find_last()
model.load_weights(model_path, by_name=True)

dataset_val = coco.CocoDataset()
dataset_val.load_coco(DATA_DIR, subset="val", year="2019")
dataset_val.prepare()
#val_type = "val" if args.year in '2017' else "minival"
val_type = "val"
#coco = dataset_val.load_coco(args.dataset, val_type, year=args.year, return_coco=True, auto_download=args.download)
dataset_all = COCO(annotation_file='/path-to-my-annotation-file-where-all-my-images-are-with-labels.json')

print("Running COCO evaluation on {} images.".format(60))
evaluate_coco(model, dataset_val, dataset_all, "segm", limit=int(60))
```
The files from Mask-RCNN are in the mrcnn and mcoco folders, see my imports.

The functions `evaluate_coco` and `build_coco_results` are from the coco.py file. 
The last part where i evaluate my results are from coco.py, too, but i adjusted some things for my purpose. You can find that part at the end of coco.py.

But after that i get a result:

```
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = -1.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
Prediction time: 160.86959218978882. Average 8.043479609489442/image
Total time:  163.76599621772766
```
Does this seem to be a normal result? Is an average of 8.04 good?
